docs/blog/2019-02-04-pragmatic-lessons-from-converting-to-react-hooks/index.md

```diff
@@ -0,0 +1,37 @@
+
+The `useEffect` hook brings more clarity and organization to the side effects in your components. Class based components force you to provide your set up and tear down across three separate life cycle methods: `componentDidMount`, `componentDidUpdate`, and `componentWillUnmount`. This often results in duplication of code. A component may use many side effects which are all placed in these same lifecycle methods. The `useEffect` Hook shifts this paradigm by isolating the functionality of a given side effect in one function. Clean up for the effect is set as a return function. This colocation allows for easier mental parsing and grouping functions by their area of concern.
+
+An extra benefit of `useState` is the ability to reference the current state without additional syntax. With Hooks you can call `setCurrIndex((currIndex + 1) % slides.length)` rather than passing a function which returns the partial state object to update. This results in cleaner code, and eliminated the need for the side-loaded function for the Slideshow component. Converting the Slideshow from a render prop component to a custom Hook resulted in a net reduction of 76 lines of code. Other components within the site utilized class components to manage a single piece of state. Hooks allow you to reduce the boilerplate of a class based component in these circumstances. Functional components connect to state with one line of code.
+
+## Structuring Hooks in Gatsby
+
+Currently I am structuring all my custom hooks in a top level folder so that other components can import them from a central location. Pulling the React dependencies out of Gatsby in version 2 allows for using Hooks immediately. To begin using Hooks today, update React and React-DOM to 16.8.0. There are considerations to take with the `useEffect Hook`. If it references the window object, you need to check that window is defined to avoid Gatsby build errors. These effects would normally live in `componentDidMount` where the component hydrates in the DOM. Hooks are called in the build process. The Gatsby docs have great resources for [debugging HTML builds](https://www.gatsbyjs.org/docs/debugging-html-builds/) if you encounter this issue.
+
+Hooks bring more options on when to convert a functional component to a class based component if at all. Custom Hooks work well for abstracting complex logic. These provide the same composable logic as render props without the added component in the tree or syntax. Hooks improve the ability to share logic across components. You can find the code conversion to Hooks discussed in this post at the [related PR](https://github.com/dslemay/Portfolio-Site/pull/10). Hooks may not be the death of class based components, but they do encourage critical thought before reaching for a class component.
```

Comment:
This is a bit wordy. One thing I've found super helpful is to break up my sentences into a list structure when concluding, e.g

```
Quite a lot of benefits of hooks are now apparent. Specifically, in this discussion, the following benefits of hooks are now clear:

- benefit a
- benefit b
- benefit c

From this basis, the value of hooks are clear and obvious. I can't wait to see what you build with hooks, and I believe they constitute a meaningful shift in the way we do React development.
```

Reply:
Updated in latest commit.